internal/reporting/markdown.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	testsBySection := testsByMarkdownSection(testResults)
+	writersBySection := map[markdownTestSection]func(*strings.Builder, []v1.Test) (bool, error){
+		flakySection:       writeMarkdownFlakySection,
+		failedSection:      writeMarkdownFailedSection,
+		timedOutSection:    writeMarkdownTimedOutSection,
+		quarantinedSection: writeMarkdownQuarantinedSection,
+		canceledSection:    writeMarkdownCanceledSection,
+	}
+
+	orderedSections := []markdownTestSection{
+		flakySection,
+		failedSection,
+		timedOutSection,
+		quarantinedSection,
+		canceledSection,
+	}
+
+	for _, section := range orderedSections {
+		shouldTruncate, err := writersBySection[section](markdown, testsBySection[section])
+		if err != nil {
+			return errors.WithStack(err)
+		}
+		if shouldTruncate {
+			if _, err := markdown.WriteString(markdownResultsTruncated); err != nil {
+				return errors.WithStack(err)
+			}
+			if _, err := file.Write([]byte(markdown.String())); err != nil {
+				return errors.WithStack(err)
+			}
+			return nil
+		}
+	}
+
+	if _, err := file.Write([]byte(markdown.String())); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func writeMarkdownSummaryStatus(markdown *strings.Builder, value int, singular string, plural string) error {
+	if value <= 0 {
+		return nil
+	}
+
+	if _, err := markdown.WriteString(fmt.Sprintf(", %v %v", value, pluralize(value, singular, plural))); err != nil {
+		return errors.WithStack(err)
+	}
+
+	return nil
+}
+
+func writeMarkdownSummaryLine(markdown *strings.Builder, testResults v1.TestResults) error {
+	summary := testResults.Summary
+
+	if _, err := markdown.WriteString(
+		fmt.Sprintf("%v %v", summary.Tests, pluralize(summary.Tests, "test", "tests")),
+	); err != nil {
+		return errors.WithStack(err)
+	}
+
+	flaky := 0
+	for _, test := range testResults.Tests {
+		if test.Flaky() {
+			flaky++
+		}
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, flaky, "flaky", "flaky"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Failed, "failed", "failed"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.TimedOut, "timed out", "timed out"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Quarantined, "quarantined", "quarantined"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Canceled, "canceled", "canceled"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Pended, "pended", "pended"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Skipped, "skipped", "skipped"); err != nil {
+		return errors.WithStack(err)
+	}
+
+	if err := writeMarkdownSummaryStatus(markdown, summary.Todo, "todo", "todo"); err != nil {
+		return errors.WithStack(err)
+	}
```

Comment:
We summarize these in the summary line, but we don't have separate sections for them (as they're not all that interesting)